src/container/srv_target.c

```diff
@@ -279,38 +485,165 @@ ds_cont_child_cache_destroy(struct daos_lru_cache *cache)
 }
 
 /*
- * If "pool == NULL", then this is assumed to be a pure lookup. In this case,
+ * If "po_uuid == NULL", then this is assumed to be a pure lookup. In this case,
  * -DER_NONEXIST is returned if the ds_cont_child object does not exist.
  */
 static int
-cont_child_lookup(struct daos_lru_cache *cache, const uuid_t uuid,
-		  struct ds_pool_child *pool, struct ds_cont_child **cont)
+cont_child_lookup(struct daos_lru_cache *cache, const uuid_t co_uuid,
+		  const uuid_t po_uuid, struct ds_cont_child **cont)
 {
 	struct daos_llink      *llink;
 	int			rc;
 
-	rc = daos_lru_ref_hold(cache, (void *)uuid, sizeof(uuid_t), pool,
-			       &llink);
+	rc = daos_lru_ref_hold(cache, (void *)co_uuid, sizeof(uuid_t),
+			       (void *)po_uuid, &llink);
 	if (rc != 0) {
 		if (rc == -DER_NONEXIST)
 			D_DEBUG(DF_DSMS, DF_CONT": failed to lookup%s "
-				"container: %d\n", DP_CONT(NULL, uuid),
-				pool == NULL ? "" : "/create", rc);
+				"container: %d\n", DP_CONT(po_uuid, co_uuid),
+				po_uuid == NULL ? "" : "/create", rc);
 		else
 			D_ERROR(DF_CONT": failed to lookup%s container: %d\n",
-				DP_CONT(NULL, uuid),
-				pool == NULL ? "" : "/create", rc);
+				DP_CONT(po_uuid, co_uuid),
+				po_uuid == NULL ? "" : "/create", rc);
 		return rc;
 	}
 
-	*cont = cont_child_obj(llink);
-	return 0;
+	*cont = cont_child_obj(llink);
+	return 0;
+}
+
+static void
+cont_child_put(struct daos_lru_cache *cache, struct ds_cont_child *cont)
+{
+	daos_lru_ref_release(cache, &cont->sc_list);
+}
+
+static inline bool
+cont_child_started(struct ds_cont_child *cont_child)
+{
+	/* Started container is linked in spc_cont_list */
+	return !d_list_empty(&cont_child->sc_link);
+}
+
+static void
+cont_child_stop(struct ds_cont_child *cont_child)
+{
+	if (!cont_child->sc_stopping) {
+		D_DEBUG(DF_DSMS, DF_CONT"[%d]: Stopping container\n",
+			DP_CONT(cont_child->sc_pool->spc_uuid,
+				cont_child->sc_uuid),
+			dss_get_module_info()->dmi_tgt_id);
+
+		cont_child->sc_stopping = 1;
+		d_list_del_init(&cont_child->sc_link);
+
+		/* cont_stop_agg_ult() may yield */
+		cont_stop_agg_ult(cont_child);
+		ds_cont_child_put(cont_child);
+	} else {
+		D_ASSERT(!cont_child_started(cont_child));
+	}
+}
+
+void
+ds_cont_child_stop_all(struct ds_pool_child *pool_child)
+{
+	d_list_t		*cont_list;
+	struct ds_cont_child	*cont_child;
+
+	D_DEBUG(DF_DSMS, DF_UUID"[%d]: Stopping all containers\n",
+		DP_UUID(pool_child->spc_uuid),
+		dss_get_module_info()->dmi_tgt_id);
+
+	D_ASSERT(d_list_empty(&pool_child->spc_list));
+
+	cont_list = &pool_child->spc_cont_list;
+	while (!d_list_empty(cont_list)) {
+		cont_child = d_list_entry(cont_list->next,
+					  struct ds_cont_child, sc_link);
+		cont_child_stop(cont_child);
```

Comment:
This will stop all containers aggregate ult sequentially?   Maybe we should notify all ULT all together, then wait?

Reply:
yeah, I think that can be improved, but not in this patch.